src/raftstore/store/store.rs

```diff
@@ -1039,16 +1173,20 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             self.raft_metrics.message_dropped.region_no_peer += 1;
             return Ok(Some(key));
         }
-        if let Some((_, &exist_region_id)) = self.region_ranges
+
+        let r = self.region_ranges
             .range((Excluded(enc_start_key(&snap_region)), Unbounded::<Key>))
+            .map(|(_, &region_id)| self.region_peers[&region_id].region())
+            .take_while(|r| enc_start_key(r) < enc_end_key(&snap_region))
+            .skip_while(|r| r.get_id() == region_id)
             .next()
-        {
-            let exist_region = self.region_peers[&exist_region_id].region();
-            if enc_start_key(exist_region) < enc_end_key(&snap_region) {
-                info!("region overlapped {:?}, {:?}", exist_region, snap_region);
-                self.raft_metrics.message_dropped.region_overlap += 1;
-                return Ok(Some(key));
-            }
+            .map(|r| r.to_owned());
+        if let Some(exist_region) = r {
+            info!("region overlapped {:?}, {:?}", exist_region, snap_region);
+            self.pending_cross_snap
+                .insert(region_id, snap_region.get_region_epoch().to_owned());
+            self.raft_metrics.message_dropped.region_overlap += 1;
+            return Ok(Some(key));
```

Comment:
Could we keep the `snapshot` for using it later?

Reply:
Generally, leader won't send same snapshot twice unless there is no write.